src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,321 @@
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <config.h>
+#include <filesystem>
+#include <array>
+#include <sstream>
+#include <inttypes.h>
+#include <sys/stat.h>
+#include <libvmi/libvmi.h>
+
+#include "ipt.h"
+#include "plugins/output_format.h"
+#include "private.h"
+
+namespace
+{
+
+uint64_t pack_payload(uint32_t cmd, uint32_t data)
+{
+    return (static_cast<uint64_t>(cmd) << 32) | data;
+}
+
+void emit_ptwrite64(std::ofstream& stream, uint64_t payload)
+{
+    // ptwrite packet
+    stream.put(0x02);
+    // no FUP, 8 bytes of payload
+    stream.put(0x32);
+    stream.write(reinterpret_cast<char*>(&payload), sizeof(payload));
+}
+
+
+event_response_t ipt_cr3_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = get_trap_plugin<ipt>(info);
+    auto& vcpu = plugin->vcpus[info->vcpu];
+
+    vcpu.flush(info->regs->vmtrace_pos);
+
+    vcpu.annotate(pack_payload(PTW_CURRENT_CR3, info->regs->cr3));
+    vcpu.annotate(pack_payload(PTW_CURRENT_TID, info->proc_data.tid));
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+event_response_t ipt_catchall_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = get_trap_plugin<ipt>(info);
+    auto& vcpu = plugin->vcpus[info->vcpu];
+
+    vcpu.flush(info->regs->vmtrace_pos);
+
+    vcpu.annotate(pack_payload(PTW_EVENT_ID, info->event_uid));
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+} // unnamed namespace
+
+void ipt_vcpu::annotate(uint64_t payload)
+{
+    emit_ptwrite64(this->output_stream, payload);
+}
+
+void ipt_vcpu::flush(uint64_t offset)
+{
+    // update last offset
+    uint64_t prev = this->last_offset;
+    this->last_offset = offset;
+
+    using ull = unsigned long long;
```

Comment:
Honestly when dealing with low level stuff where size of the type really maters I prefer to use `uint64_t`. At least you see why it is used and what size it is without deciphering what `unsigned long long long long long` means.